UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c

```diff
@@ -226,31 +225,29 @@ SmmInitPageTable (
 
   //
   // Generate initial SMM page table.
-  // Only map [0, 4G] when PcdCpuSmmRestrictedMemoryAccess is FALSE.
   //
-  PhysicalAddressBits = mCpuSmmRestrictedMemoryAccess ? mPhysicalAddressBits : 32;
-  PageTable           = GenSmmPageTable (mPagingMode, PhysicalAddressBits);
+  PageTable = GenSmmPageTable (mPagingMode, mPhysicalAddressBits);
+
+  if (FeaturePcdGet (PcdCpuSmmProfileEnable)) {
+    if (m5LevelPagingNeeded) {
+      Pml5Entry = (UINT64 *)PageTable;
+      //
+      // Set Pml5Entry sub-entries number for smm PF handler usage.
+      //
+      SetSubEntriesNum (Pml5Entry, 1);

```

Comment:
It's for dynamic creating page table.

"do we understand why original code sets the Pml5Entry SubEntries to 2?"
I think it's not needed when mCpuSmmRestrictedMemoryAccess is TRUE.

For now, it's only needed when SMM profile is enabled

Reply:
could you explain why need 2 sub-entries of Pml5Entry for dynamic creating page table?
SetSubEntriesNum (Pml5Entry, 1) means set the SubEntries to 2.

Reply:
It's related with the page reclaim mechanism. If the subentry num of a non-leaf PDPDE is 0, when the child of this non-leaf entry is reclaimed, the non-leaf parent may also be relaimed since the only present child is reclaimed.

At first, I thought that it's set to non-zero value to prevent the PML4 paging structure pointered by PLM5[0] from being reclaimed in the page reclaim mechanism. This is to make sure [0, 4G] is always present in SMM page table

But I checked code and found that there is no logic to reclaim any PML4 buffer pointered by a PML5 entry. So I think this line of code to set subentry for PML5 is not needed. But it's not related with the task and doesn't impact anything. I think we can move it in our future plan that we may totally replace the page reclaim mechanism with a new and simpler mechanism